R/method-correlation.R

```diff
@@ -0,0 +1,173 @@
+
+    # Check that incoming abundance data meets requirements
+    if (!'data.table' %in% class(df1)) {
+      data.table::setDT(df1)
+    }
+
+    # Check that incoming metadata meets requirements
+    if (!'data.table' %in% class(df2)) {
+      data.table::setDT(df2)
+    }
+
+    # Check that variables is the correct class
+    if (!is.null(variables) && !'VariableMetadataList' %in% class(variables)) {
+      stop("The variables argument must be a VariableMetadataList if defined.")
+    }
+
+    computeMessage <- ''
+    veupathUtils::logWithTime(paste("Received df table with", nrow(df1), "samples and", (ncol(df1)-1), "taxa."), verbose)
+
+    # Replace NA values with 0
+    if (naToZero) {
+      veupathUtils::setNaToZero(df1)
+      veupathUtils::logWithTime("Replaced NAs with 0.", verbose)
+    }
+
+    # Ensure the sample IDs, and their order, are the same in both datasets
+    if (!identical(df1[[recordIdColumn1]], df2[[recordIdColumn2]])) {
+      if (!identical(union(df1[[recordIdColumn1]], df2[[recordIdColumn2]]), intersect(df1[[recordIdColumn1]], df2[[recordIdColumn2]]))) {
+        stop("Sample IDs do not match between data1 and data2.")
+      }
+      # Reorder df2 (sample metadata) so that it matches the order in df
+      sampleMetadata <- df2[match(df1[[recordIdColumn1]], df2[[recordIdColumn2]]), ]
+      veupathUtils::logWithTime("Reordered sampleMetadata rows based on abundance data sample id order.", verbose)
+    }
+
+    ## Extract appropriate metadata columns
+    if (!is.null(variables)) {
+      # Use the data shape to find appropriate columns in the metadata
+      inputMetadataCols <- veupathUtils::findColNamesByPredicate(variables, function(x) {identical(x@dataShape@value, "CONTINUOUS")})
+
+      # If any are dates, we'll need to coerce them to numeric for the correlation
+      dateMetadataCols <- veupathUtils::findColNamesByPredicate(variables, function(x) {identical(x@dataShape@value, "CONTINUOUS") && identical(x@dataType@value, "DATE")})
+      if (!!length(dateMetadataCols)) {
+        veupathUtils::logWithTime("Converting date variables to numeric", verbose)
+        df2[, (dateMetadataCols) := lapply(.SD, as.numeric), .SDcols = dateMetadataCols]
+      }
+
+    } else {
+      warning("No variable metadata supplied. Using all numeric columns of sampleMetadata for correlation.")
+      inputMetadataCols <- veupathUtils::findNumericCols(df2[, -..allIdColumns2])
+    }
+    if (length(inputMetadataCols) < 1) {
+      stop("correlation requires at least one continuous metadata variable.")
+    }
+
+
+    ## Compute correlation
+    # Result has columns "data1", "data2", and correlationCoef
+    corrResult <- correlation(df1[, -..allIdColumns1], df2[, ..inputMetadataCols], method = method, verbose = verbose)
+
+
+    ## Format results
+    # Rename the correlation results columns so that they are a little more friendly
+    names(corrResult)[names(corrResult) == "data1"] <- data1Name
```

Comment:
not sure i like this. wondering if we can have a CorrelationResult class or something, that has a slot for the data.frame your default method returns, but also slots for things like 'data1Metadata` and `data2Metadata` where we could say that data1 is sample data or assay data etc.

Reply:
also, why are these params of the method? doesnt the signature here guarantee were working w assay v metadata? am i misunderstanding what the names are for?

Reply:
i understood the goal to be to let us more specifically access the data, so if we only know the classes we'll run into trouble when we have assay v assay. In that case we'd have to basically name it assay1 and assay2 if the method just had to define these names itself. So I was thinking we'd pass in the collection ID to dataName1 and "metadata" or whatever into dataName2. 

But i think i probably misunderstood the original use for this type of thing. Because even if we have a CorrelationResult class then for assay v assay  it'd also have "Assay" in both dataYMetadata slots?

Reply:
I think the goal ultimately is to know what data1 and data2 represent. We want to be able to reference them generically I think, so the component+ viz for ex can just say put whatever 'data1' is over here or there. But also it needs to know what data1 fundamentally IS to do things like put display labels places? And/ or validate data1 is what its meant to be too maybe? Maybe the metadata slot is a VariableMetadata object?

Reply:
got it. Okay that makes sense!